### PR TITLE
Instance segmentation support with the new Datumaro dataset

### PR DESCRIPTION
**Pull request by gdlg**
### Summary

This PR adds support for the new Datumaro dataset for instance segmentation. It also refactors the handling of polygon masks for instance segmentation tasks, transitioning from the use of `datumaro.Polygon` objects to a ragged array `np.ndarray` format. The changes propagate this new representation throughout the data pipeline, including dataset classes, sample entities, validation logic, and transformation utilities.

**Polygon mask representation update:**

* Replaced all usages of `datumaro.Polygon` with ragged `np.ndarray` arrays for polygon masks in dataset entities, validation logic, and model inputs/outputs, including updates to type annotations and validators in files like `instance_segmentation.py`, `sample.py`, and `torch.py`. [[1]](diffhunk://#diff-5879f23f99c01a3859cbaba3facd22410dfd3f25cb9de99c366b736e338060cbR24-R45) [[2]](diffhunk://#diff-352798ef671935a54be18e1593c47a39afadb474419360b5d33663294ee334efL58-R66) [[3]](diffhunk://#diff-1a83603f4abcd216f809367ee905848bd8292b6f3e34848f2c3612c92ad98d53L53-R53) [[4]](diffhunk://#diff-777fbd7dbd6eaea5371c6c8e06a20f3ca2bf48f6a0339544a5d20d7981936ea3L157-R168) [[5]](diffhunk://#diff-777fbd7dbd6eaea5371c6c8e06a20f3ca2bf48f6a0339544a5d20d7981936ea3L398-R408) [[6]](diffhunk://#diff-c017d7350917f2ea72e7674caaad3abdf16315ec3e5e606edc5407ed5d8debb2L18-R24) [[7]](diffhunk://#diff-c017d7350917f2ea72e7674caaad3abdf16315ec3e5e606edc5407ed5d8debb2L34-R34) [[8]](diffhunk://#diff-d25d847b4429c24ec16d0c9014291899e058f22fcaedf73df10f41c454c2a95dL647-R646)
* Added a utility function `convert_datumaro_polygons_to_ragged_array` to convert lists of `datumaro.Polygon` objects into the new ragged array format.

**Instance segmentation dataset and sample improvements:**

* Introduced a new `OTXInstanceSegDataset` class in `instance_segmentation_new.py` and updated the data factory to use this for instance segmentation and rotated detection tasks, ensuring compatibility with the new polygon format. [[1]](diffhunk://#diff-a7b15374fe3d716a5f6a7e9ccbc5df62f919a597469da4f7bf6cbd8cd810a23bR1-R50) [[2]](diffhunk://#diff-535a232efb8b265495f79f79fe09b2dc476c6cf5b18efb25c176ba0550d9c015L121-R127)
* Added the `InstanceSegmentationSample` entity class, supporting ragged polygon arrays and proper conversion to PyTorch tensor formats.

**Transformation and augmentation updates:**

* Refactored polygon-related transformations to operate on ragged arrays, including filtering, projection, and bounding box recomputation in `torchvision.py`. [[1]](diffhunk://#diff-272183479d12d6e1939a849c2b95b97b774c362e299a113cacbf5b3d3a8bdbfaL1412-R1413) [[2]](diffhunk://#diff-272183479d12d6e1939a849c2b95b97b774c362e299a113cacbf5b3d3a8bdbfaL1445-R1448) [[3]](diffhunk://#diff-272183479d12d6e1939a849c2b95b97b774c362e299a113cacbf5b3d3a8bdbfaL1768-R1766)

**Model and loss function adjustments:**

* Updated model heads and loss functions to accept and process polygon masks as ragged arrays (`np.ndarray`), removing dependencies on `datumaro.Polygon`. [[1]](diffhunk://#diff-d25d847b4429c24ec16d0c9014291899e058f22fcaedf73df10f41c454c2a95dL21) [[2]](diffhunk://#diff-c017d7350917f2ea72e7674caaad3abdf16315ec3e5e606edc5407ed5d8debb2L18-R24) [[3]](diffhunk://#diff-c017d7350917f2ea72e7674caaad3abdf16315ec3e5e606edc5407ed5d8debb2L34-R34) [[4]](diffhunk://#diff-e20a49fde79db5fe038b696c7c830caff23fe98712a6e536b7c5161a5fa340e2L61-R78)

**Miscellaneous and compatibility:**

* Minor fixes and improvements for compatibility, such as type checking, import adjustments, and mock module handling. [[1]](diffhunk://#diff-7ddb83772ee558ecd5accb0192b5ab8b86cde4d3679327ae82ea73d8c17fe033L83-R84) [[2]](diffhunk://#diff-272183479d12d6e1939a849c2b95b97b774c362e299a113cacbf5b3d3a8bdbfaL10) [[3]](diffhunk://#diff-272183479d12d6e1939a849c2b95b97b774c362e299a113cacbf5b3d3a8bdbfaR38) [[4]](diffhunk://#diff-535a232efb8b265495f79f79fe09b2dc476c6cf5b18efb25c176ba0550d9c015L56-R57)

Let me know if you want to walk through any specific part of this refactor or need help understanding how to work with the new polygon format in your code!

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have ran e2e tests and there is no issues.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/open-edge-platform/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/open-edge-platform/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/open-edge-platform/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2025 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: lib/src/otx/backend/native/models/instance_segmentation/rotated_det.py [hunk 1]
```diff
@@ -4,13 +4,27 @@
...
 import cv2
+import numpy as np
 import torch
-from datumaro import Polygon
 from torchvision import tv_tensors
 
 from otx.data.entity.torch.torch import OTXPredBatch
 
 
+def get_polygon_area(points: np.ndarray) -> float:
```
Comment (gdlg) [inline lib/src/otx/backend/native/models/instance_segmentation/rotated_det.py:14]: We can consider moving this function later to Datumaro when the new Datumaro API is a bit more stable.

File: lib/src/otx/data/transform_libs/torchvision.py [hunk 2]
```diff
@@ -2040,7 +2036,7 @@ def forward(self, *_inputs: OTXDataItem) -> OTXDataItem | None:
         mixup_img = 0.5 * ori_img + 0.5 * padded_cropped_img.astype(np.float32)
 
         # TODO(ashwinvaidya17): remove this once we have a unified TorchDataItem
-        if isinstance(retrieve_results, OTXDataItem):
+        if isinstance(retrieve_results, (OTXDataItem, OTXSample)):
```
Comment (gdlg) [inline lib/src/otx/data/transform_libs/torchvision.py:2039]: I am leaving this for now. Once we have finished migrating the datasets to the new Datumaro API, I would like to simplify and unify the data structures used by the models.

File: lib/src/otx/data/utils/structures/mask/mask_util.py [hunk 3]
```diff
@@ -10,52 +10,66 @@
 import numpy as np
 import pycocotools.mask as mask_utils
 import torch
-from datumaro import Polygon
 from torchvision.ops import roi_align
 
 if TYPE_CHECKING:
+    from datumaro import Polygon
```
Comment (eugene123tw) [inline lib/src/otx/data/utils/structures/mask/mask_util.py]: Is there a reason we keep datumaro.Polygon for `polygon_to_bitmap` and `polygon_to_rle`?

**Review by eugene123tw — COMMENTED** [review]
Thanks Greg for the efforts! I left one comment, the rest looks good! Thank you.

**Review by AlbertvanHouten — APPROVED** [review]

File: lib/src/otx/data/utils/structures/mask/mask_util.py (hunk 3 above)
Comment (gdlg) [reply lib/src/otx/data/utils/structures/mask/mask_util.py]: Thanks for spotting this. It’s a leftover and I have removed it now.

File: lib/pyproject.toml
Comment (AlbertvanHouten) [inline lib/pyproject.toml:1]: Seems like the changes in this file should be reverted

Comment (gdlg) [reply lib/pyproject.toml:1]: Fixed!

**Review by AlbertvanHouten — APPROVED** [review]
